apis/placement/v1alpha1/stagerollout_types.go

```diff
@@ -0,0 +1,344 @@
+// +enum
+type PostRolloutTaskConditionType string
+
+const (
+	// PostRolloutTaskConditionApprovalRequestCreated indicates whether the approval request is created.
+	// Its condition status can be one of the following:
+	// - "True" means the approval request is created.
+	// - "False" means the approval request is not created.
+	PostRolloutTaskConditionApprovalRequestCreated PostRolloutTaskConditionType = "ApprovalRequestCreated"
+
+	// PostRolloutTaskConditionApprovalRequestApproved indicates whether the approval request is approved.
+	// Its condition status can be one of the following:
+	// - "True" means the approval request is approved.
+	// - "False" means the approval request is not approved.
+	PostRolloutTaskConditionApprovalRequestApproved PostRolloutTaskConditionType = "ApprovalRequestApproved"
+
+	// PostRolloutTaskConditionApprovalWaitTimeElapsed indicates whether the wait time after each stage is elapsed.
+	// We will fill the message of the condition of the remaining wait time if the status is "False".
+	// Its condition status can be one of the following:
+	// - "True" means the  wait time is elapsed.
+	// - "False" means the  wait time is not elapsed.
+	PostRolloutTaskConditionApprovalWaitTimeElapsed PostRolloutTaskConditionType = "WaitTimeElapsed"
+)
+
+// StageRolloutConditionType identifies a specific condition of the StagedRolloutRun.
+// +enum
+type StageRolloutConditionType string
+
+const (
+	// StageRolloutConditionRollingOut indicates whether the stage is rolling out normally.
+	// Its condition status can be one of the following:
+	// - "True" means the stage is rolling out.
+	// - "False" means the stage rolling out is not progressing.
+	// - "Unknown" means it is unknown.
+	StageRolloutConditionRollingOut StageRolloutConditionType = "RollingOut"
+
+	// TODO: add alerting condition
+)
+
+// StageRolloutList contains a list of StagedRolloutRun.
+// +kubebuilder:resource:scope="Cluster"
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type StageRolloutList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []StagedRolloutRun `json:"items"`
+}
+
+// +genclient
+// +genclient:nonNamespaced
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:scope="Namespaced",categories={fleet,fleet-placement}
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// RolloutApprovalRequest defines a stage by stage rollout policy that is applied to the ClusterResourcePlacement of the same name.
+// Each snapshot MUST have the following labels:
+//   - `ParentStageRollout` which points to its owner stage rollout.
+//   - `TargetStage` which is the name of the stage that this approval request is for.
+//   - `IsLatestRolloutApproval` which indicates whether this approval request is the latest one related to this rollout
+type RolloutApprovalRequest struct {
```

Comment:
Which namespace should the rollout approve request be in? Will it be in the same one as rollout update run? How do we manage access control of approve requests vs runs vs strategies?